### ISSUE TITLE
Implicit conversion of scalars to tensors

### ISSUE THREAD

**Issue by MovGP0** [dotnet/TorchSharp#566]
There should be an implicit conversion of common number types to scalar tensors. Instead of:

```csharp
var points = torch.zeros(6, torch.ScalarType.Float16);
points[0] = torch.tensor(4.0f, torch.ScalarType.Float16);
points[1] = torch.tensor(1.0d, torch.ScalarType.Float32);
points[2] = torch.tensor(7, torch.ScalarType.Int32);
```

we should be able to write

```csharp
var points = torch.zeros(6, torch.ScalarType.Float16);
points[0] = 4.0f;
points[1] = 1.0d;
points[2] = 7;
```

**Comment by NiklasGustafsson** [comment]
@MovGP0 -- that was the intent at one point, but it created a lot of ambiguity with existing APIs with overloads that took both a tensor and a scalar.

There is an explicit conversion, so you should be able to say:

```C#
var points = torch.zeros(6, torch.ScalarType.Float16);
points[0] = (Tensor)4.0f;
points[1] = (Tensor)1.0d;
points[2] = (Tensor)7;
```

which is slightly less cumbersome.

**Comment by MovGP0** [comment]
There is the convention that converting without Information loss should be implicit and converting with Information reduction should be explicit.

The conversion from int/float/double/etc. to Tensor would increase Information and can be implicit, while the conversion from a scalar Tensor to a numeric value would need to be explicit.

Ambiguity shouldn't be a problem in theory. If there is ambiguity, it's usually  better to remove the interfaces that use a primitive type in favor of interfaces that use the Tensor, since there will be an automatic conversion anyway.

**Comment by NiklasGustafsson** [comment]
Hmm, I'm going to have to think about that one. 

I've always thought of that convention in terms of "widening" or "narrowing" conversions, but to implicitly create a collection type from a scalar seems different from widening. I'm not immediately persuaded that it meets the criteria. 

That said, TorchSharp breaks many .NET conventions already, so I wouldn't personally let that stand in the way of usability. It's been a while since I looked at what started breaking when those operators were implicit, and I'll take another look.

**Comment by NiklasGustafsson** [comment]
The specific challenge with indexing is that we can't overload the setters on the value type. Otherwise, we could fix this by taking 'Scalar':

```C#
[IndexerName("TensorItems")]
public Tensor this[long i1] {
    get { ... }
    set { ... }
}

[IndexerName("TensorItems")]
public Scalar this[long i1] {
    set { ... }
}
```
but that's not legal C#.

A possibility (but it feels like a hack) is to invent a type TensorOrScalar, derived from Tensor. The Tensor indexer would then take TensorOrScalar as its value type. Since TensorOrScalar wouldn't conflict with Scalar in any places where it's being used, it **should** work, at the cost of feeling a bit kludgy.

**Comment by MovGP0** [comment]
I can think of two ways of solving the indexer example. Either use a common base class or interface for `Scalar` and `Tensor` 

```csharp
[IndexerName("TensorItems")]
public ITensor this[long i1] =>... 
```
or a Discriminated Union
```csharp
[IndexerName("TensorItems")]
public Union<Tensor, Scalar> this[long i1] =>... 
```
The latter is native in F#
```fsharp
type TensorOrScalar = T of Tensor | S of Scalar
```

The discriminated union would be cleaner from a theoretical standpoint, since it forces the user to handle all possibilities, but would also break existing logic.

**Comment by NiklasGustafsson** [comment]
The unfortunate bit is that, for any one of these solutions, the getter and the setters cannot have distinct signatures, so while the setter could take a "combined" type, the getter would have to return the same type, and the calling code would have to discriminate.

```C#
var x = t[i];
```

the static type of `x` would be Union<Tensor, Scalar> (or ITensor), not Tensor, even though it's always going to be a Tensor instance.

**Comment by NiklasGustafsson** [comment]
So, the only way I can think of making this work, is to make the operator implicit and then remove the various arithmetic overloads that involve "Scalar" and replace them with overloads that use all the arithmetic scalar types. That is, what you suggested earlier.

That's going to be a lot of monkey-typing / cut-pasting.

**Comment by NiklasGustafsson** [comment]
@MovGP0, 

There's good news here. According to C# master @MadsTorgersen, the trick is to make all the conversion operators implicit, and then also add an implicit operator Scalar -> Tensor. That will cause the compiler to choose the right overloads for things like `t + 4711.5`, but also allow the direct assignment of .NET scalars to tensor elements.

The Scalar -> Tensor operator should never be used, just be there to guide the compiler:

```C#
public static implicit operator Tensor(Scalar scalar)
{
    throw new InvalidOperationException("Implicit conversion from Scalar to Tensor.");
}
```

With that in place, everything just works, including with the latest F# compiler.

I'll PR this as soon as I've made a few unit tests.

**Comment by NiklasGustafsson** [comment]
Addressed in PR #574
